lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,213 @@
+require 'optparse'
+require 'openssl'
+require 'net/https'
+require 'json'
+
+module Puppetserver
+  module Ca
+    class SignAction
+
+      SUMMARY = "Set up the CA's key, certs, and crls"
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca sign [--help|--version]
+  puppetserver ca sign  --certname <comma separated list of hostnames>
+  puppetserver ca sign  --all
+
+Description:
+Given a comma-separated list of valid certnames, instructs the CA to sign each cert.
+
+Options:
+      BANNER
+      BODY = JSON.dump({desired_state: 'signed'})
+      HEADERS = {
+          'User-Agent' => 'PuppetserverCaCli',
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json'}
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--certname x,y,z', Array, 'the name(s) of the cert(s) to be signed') do |cert|
+            parsed['certname'] = cert
+          end
+          opts.on('--help', 'Display this sign specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'Operate on all certnames') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        if input['all']
+          requested_certnames = get_all_pending_certs
+          if requested_certnames.nil?
+            return 1
+          end
+        else
+          requested_certnames = input['certname']
+        end
+
+        exit_code = sign_requested_certs(requested_certnames)
+        return exit_code
+      end
+
+      def get_certificate_statuses
+        uri = URI('https://localhost:8140/puppet-ca/v1/certificate_statuses/:any_key?environment=production')
```

Comment:
Is there no way to explicitly request certs with state of "requested" ?

Reply:
I don't think so, or at least, I didn't see anything about it [here](https://puppet.com/docs/puppet/5.3/http_api/http_certificate_status.html).

Reply:
Sadly, [no](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/puppetserver/certificate_authority.clj#L1210-L1235).

I'm not finding the ticket but there's a feature request for server side filtering of statuses which would be great as an optimization for this afterwards.